jlm/llvm/opt/alias-analyses/AliasAnalysisPrecisionEvaluator.cpp

```diff
@@ -0,0 +1,501 @@
+  {
+    auto [p1, s1, p1IsClobber, p1Multiplier] = Context_.PointerOperations[i];
+
+    precisionEvaluation.NumOperations += p1Multiplier;
+
+    if (!p1IsClobber)
+      continue;
+
+    precisionEvaluation.NumClobberOperations += p1Multiplier;
+
+    PrecisionInfo::ClobberInfo clobberInfo;
+    clobberInfo.Multiplier = p1Multiplier;
+
+    // If p1 represents more than one concrete clobber operation,
+    // add the result of querying against all the other clobber operations
+    if (p1Multiplier > 1)
+      clobberInfo.NumMustAlias += p1Multiplier - 1;
+
+    for (size_t j = 0; j < Context_.PointerOperations.size(); j++)
+    {
+      if (i == j)
+        continue;
+
+      auto [p2, s2, p2IsClobber, p2Multiplier] = Context_.PointerOperations[j];
+
+      auto response = aliasAnalysis.Query(*p1, s1, *p2, s2);
+
+      // Queries should always be symmetric, so double check that in debug builds
+      // Note: LLVM's own alias analyses are not always symmetric, but ours should be
+      JLM_ASSERT(response == aliasAnalysis.Query(*p2, s2, *p1, s1));
+
+      // Add edge to aliasing graph if dumping a graph of alias analysis response edges is requested
+      if (IsAliasingGraphEnabled())
+      {
+        AddToAliasingGraph(*p1, s1, *p2, s2, response);
+      }
+
+      if (response == AliasAnalysis::NoAlias)
+        clobberInfo.NumNoAlias += p2Multiplier;
+      else if (response == AliasAnalysis::MayAlias)
+        clobberInfo.NumMayAlias += p2Multiplier;
+      else if (response == AliasAnalysis::MustAlias)
+        clobberInfo.NumMustAlias += p2Multiplier;
+      else
+        JLM_UNREACHABLE("Unknown AliasAnalysis query response");
+    }
+
+    precisionEvaluation.ClobberOperations.push_back(clobberInfo);
+  }
+}
+
+void
+AliasAnalysisPrecisionEvaluator::CollectPointersFromRegion(const rvsdg::Region & region)
+{
+  for (auto & node : region.Nodes())
+  {
+    if (auto simpleNode = dynamic_cast<const rvsdg::SimpleNode *>(&node))
+    {
+      CollectPointersFromSimpleNode(*simpleNode);
+    }
+    else if (auto structuralNode = dynamic_cast<const rvsdg::StructuralNode *>(&node))
+    {
+      CollectPointersFromStructuralNode(*structuralNode);
+    }
+    else
+    {
+      JLM_UNREACHABLE("Unknown node type");
+    }
+  }
+}
+
+void
+AliasAnalysisPrecisionEvaluator::CollectPointersFromSimpleNode(const rvsdg::SimpleNode & node)
+{
+  // In this mode, only (volatile) load and store operations count as uses and clobbers
+  if (const auto load = dynamic_cast<const LoadOperation *>(&node.GetOperation()))
+  {
+    const auto size = GetTypeSize(*load->GetLoadedType());
+    CollectPointer(LoadOperation::AddressInput(node).origin(), size, AreLoadsConsideredClobbers());
+  }
+  else if (auto store = dynamic_cast<const StoreOperation *>(&node.GetOperation()))
+  {
+    const auto size = GetTypeSize(store->GetStoredType());
+    CollectPointer(StoreOperation::AddressInput(node).origin(), size, true);
+  }
```

Comment:
This code will silently change if we ever get another subclass on the LoadOperation or StoreOperation. I would rather have you spelling it out for all the individual nodes and then having an unreachable here. Not saying the Load/Store-Operation subclasses will ever get more, but we are not done modeling everything from LLVM yet.

Reply:
I added asserts to yell at us if we ever start adding new subclasses of `LoadOperation` or `StoreOperation`.